File: Python/dynamic-programming/best_time_to_buy_and_sell_stock.py
def maxProfit(prices):
    """
    :type prices: List[int]
    :rtype: int
    """
    length = len(prices)

    if length ==0:
        return 0

    profit = 0
    bottom = prices[0]
    for i in range(0, length):
        profit = max(profit, prices[i] - bottom)
        if bottom > prices[i]:
            bottom = prices[i]
    
    return profit

if __name__ == "__main__":
    print(maxProfit(prices=[7,1,5,3,6,4]))

File: Java/src/com/dbc/algorithm/DynamicProgramming/BestTimeToBuyAndSellStock.java
package com.dbc.algorithm.DynamicProgramming;

public class BestTimeToBuyAndSellStock {
    public static int maxProfit(int[] prices) {
        int len = prices.length;

        if (len == 0)
            return 0;

        int bottom = prices[0];
        int profit = 0;
        for (int i = 0; i < len; i++) {
            profit = Math.max(profit, prices[i] - bottom);
            bottom = Math.min(bottom, prices[i]);
        }

        return profit;
    }
}
